File: python/vedadb_orm/validators.py
"""
VedaDB ORM — Field validators.

Each public function returns a *callable* that accepts a single value and
raises ``ValidationError`` when the constraint is violated.  Validators are
attached to ``Field`` definitions and executed by the validation hook before
CREATE / UPDATE.
"""

import re
from typing import Any, Callable, Optional, Sequence

from .exceptions import ValidationError


# ---------------------------------------------------------------------------
# Validator factories
# ---------------------------------------------------------------------------

def required(value: Any) -> None:
    """Raise ``ValidationError`` if *value* is ``None`` or empty string."""
    if value is None or (isinstance(value, str) and value.strip() == ""):
        raise ValidationError("This field is required")


def min_length(n: int) -> Callable[[Any], None]:
    """Return a validator that enforces ``len(value) >= n``."""

    def _check(value: Any) -> None:
        if value is not None and len(str(value)) < n:
            raise ValidationError(f"Minimum length is {n}")

    return _check


def max_length(n: int) -> Callable[[Any], None]:
    """Return a validator that enforces ``len(value) <= n``."""

    def _check(value: Any) -> None:
        if value is not None and len(str(value)) > n:
            raise ValidationError(f"Maximum length is {n}")

    return _check


def min_value(n: float) -> Callable[[Any], None]:
    """Return a validator that enforces ``value >= n``."""

    def _check(value: Any) -> None:
        if value is not None and float(value) < n:
            raise ValidationError(f"Minimum value is {n}")

    return _check


def max_value(n: float) -> Callable[[Any], None]:
    """Return a validator that enforces ``value <= n``."""

    def _check(value: Any) -> None:
        if value is not None and float(value) > n:
            raise ValidationError(f"Maximum value is {n}")

    return _check


def regex(pattern: str) -> Callable[[Any], None]:
    """Return a validator that checks *value* against a regex *pattern*."""
    compiled = re.compile(pattern)

    def _check(value: Any) -> None:
        if value is not None and not compiled.match(str(value)):
            raise ValidationError(f"Value does not match pattern {pattern!r}")

    return _check


def email() -> Callable[[Any], None]:
    """Return a validator that checks for a plausible email address."""
    _EMAIL_RE = re.compile(
        r"^[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}$"
    )

    def _check(value: Any) -> None:
        if value is not None and not _EMAIL_RE.match(str(value)):
            raise ValidationError("Invalid email address")

    return _check


def one_of(*choices: Any) -> Callable[[Any], None]:
    """Return a validator that ensures *value* is among *choices*."""

    def _check(value: Any) -> None:
        if value is not None and value not in choices:
            raise ValidationError(
                f"Value must be one of {choices!r}, got {value!r}"
            )

    return _check


def custom(fn: Callable[[Any], Optional[str]]) -> Callable[[Any], None]:
    """Return a validator backed by an arbitrary function *fn*.

    *fn* should return ``None`` on success or an error-message string on
    failure.  If it raises ``ValidationError`` directly, it is re-raised
    as-is.
    """

    def _check(value: Any) -> None:
        try:
            result = fn(value)
        except ValidationError:
            raise
        except Exception as exc:
            raise ValidationError(f"Custom validation failed: {exc}") from exc
        if result is not None:
            raise ValidationError(str(result))

    return _check

File: node/vedadb-orm/src/validators.ts
/**
 * VedaDB ORM Validators
 */

export interface Validator {
  validate: (value: any) => boolean;
  message: string;
}

/**
 * Value must not be null or undefined.
 */
export function required(): Validator {
  return {
    validate: (value: any) => value !== null && value !== undefined && value !== '',
    message: 'Value is required',
  };
}

/**
 * String must be at least `n` characters long.
 */
export function minLength(n: number): Validator {
  return {
    validate: (value: any) => typeof value === 'string' && value.length >= n,
    message: `Must be at least ${n} characters`,
  };
}

/**
 * String must be at most `n` characters long.
 */
export function maxLength(n: number): Validator {
  return {
    validate: (value: any) => typeof value === 'string' && value.length <= n,
    message: `Must be at most ${n} characters`,
  };
}

/**
 * Numeric value must be at least `n`.
 */
export function minValue(n: number): Validator {
  return {
    validate: (value: any) => typeof value === 'number' && value >= n,
    message: `Must be at least ${n}`,
  };
}

/**
 * Numeric value must be at most `n`.
 */
export function maxValue(n: number): Validator {
  return {
    validate: (value: any) => typeof value === 'number' && value <= n,
    message: `Must be at most ${n}`,
  };
}

/**
 * Value must match the given regular expression.
 */
export function regex(pattern: RegExp): Validator {
  return {
    validate: (value: any) => typeof value === 'string' && pattern.test(value),
    message: `Must match pattern ${pattern}`,
  };
}

/**
 * Value must be a valid email address.
 */
export function isEmail(): Validator {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return {
    validate: (value: any) => typeof value === 'string' && emailRegex.test(value),
    message: 'Must be a valid email address',
  };
}

/**
 * Value must be one of the given choices.
 */
export function oneOf(...choices: any[]): Validator {
  return {
    validate: (value: any) => choices.includes(value),
    message: `Must be one of: ${choices.join(', ')}`,
  };
}

/**
 * Custom validator with a user-supplied function and message.
 */
export function custom(fn: (value: any) => boolean, message: string): Validator {
  return {
    validate: fn,
    message,
  };
}
